File: ios/Shared/CountryCatalog.swift
//
//  CountryCatalog.swift
//  GuideStreamTV (SHARED target — must compile into BOTH iOS and tvOS)
//
//  Curated destinations for the "Around the World" feature: region code,
//  display name, optional original-language filter, and the ordered TMDB
//  watch providers that carry titles in that region. TMDB provider ids are
//  region-specific — always use them exactly as listed here.
//
//  RORK MAX NOTE: this file lives in the SHARED group so both iOS and tvOS
//  targets compile it. It has no UIKit/SwiftUI imports.
//

import Foundation

/// One TMDB watch provider in a specific region.
nonisolated struct CountryProvider: Sendable, Equatable {
    let id: Int
    let name: String
}

/// One curated "Around the World" destination.
nonisolated struct CountryCatalogEntry: Sendable, Equatable, Identifiable {
    /// ISO 3166-1 alpha-2 region code (also TMDB's `watch_region` value).
    let regionCode: String
    let displayName: String
    /// TMDB original-language filter for local-content discovery, or nil for
    /// English-speaking markets that browse without a language filter.
    let originalLanguage: String?
    /// Ordered TMDB provider id/name pairs — region-specific ids.
    let providers: [CountryProvider]

    var id: String { regionCode }
}

nonisolated enum CountryCatalog {
    /// Curated destinations in rotation order.
    static let entries: [CountryCatalogEntry] = [
        CountryCatalogEntry(regionCode: "JP", displayName: "Japan", originalLanguage: "ja", providers: [
            CountryProvider(id: 8, name: "Netflix"),
            CountryProvider(id: 84, name: "U-NEXT"),
            CountryProvider(id: 9, name: "Amazon Prime Video"),
            CountryProvider(id: 337, name: "Disney Plus"),
            CountryProvider(id: 15, name: "Hulu"),
        ]),
        CountryCatalogEntry(regionCode: "KR", displayName: "South Korea", originalLanguage: "ko", providers: [
            CountryProvider(id: 8, name: "Netflix"),
            CountryProvider(id: 356, name: "wavve"),
            CountryProvider(id: 1883, name: "TVING"),
            CountryProvider(id: 1881, name: "Coupang Play"),
            CountryProvider(id: 97, name: "Watcha"),
            CountryProvider(id: 119, name: "Amazon Prime Video"),
        ]),
        CountryCatalogEntry(regionCode: "GB", displayName: "United Kingdom", originalLanguage: nil, providers: [
            CountryProvider(id: 8, name: "Netflix"),
            CountryProvider(id: 38, name: "BBC iPlayer"),
            CountryProvider(id: 41, name: "ITVX"),
            CountryProvider(id: 103, name: "Channel 4"),
            CountryProvider(id: 29, name: "Sky Go"),
            CountryProvider(id: 39, name: "Now TV"),
            CountryProvider(id: 9, name: "Amazon Prime Video"),
            CountryProvider(id: 337, name: "Disney Plus"),
        ]),
        CountryCatalogEntry(regionCode: "FR", displayName: "France", originalLanguage: "fr", providers: [
            CountryProvider(id: 8, name: "Netflix"),
            CountryProvider(id: 119, name: "Amazon Prime Video"),
            CountryProvider(id: 337, name: "Disney Plus"),
            CountryProvider(id: 147, name: "M6+"),
            CountryProvider(id: 531, name: "Paramount Plus"),
            CountryProvider(id: 283, name: "Crunchyroll"),
        ]),
        CountryCatalogEntry(regionCode: "DE", displayName: "Germany", originalLanguage: "de", providers: [
            CountryProvider(id: 8, name: "Netflix"),
            CountryProvider(id: 9, name: "Amazon Prime Video"),
            CountryProvider(id: 2750, name: "RTL+"),
            CountryProvider(id: 304, name: "Joyn"),
            CountryProvider(id: 30, name: "WOW"),
            CountryProvider(id: 219, name: "ARD Mediathek"),
            CountryProvider(id: 337, name: "Disney Plus"),
            CountryProvider(id: 29, name: "Sky Go"),
        ]),
        CountryCatalogEntry(regionCode: "ES", displayName: "Spain", originalLanguage: "es", providers: [
            CountryProvider(id: 8, name: "Netflix"),
            CountryProvider(id: 119, name: "Amazon Prime Video"),
            CountryProvider(id: 2241, name: "Movistar Plus+"),
            CountryProvider(id: 62, name: "Atres Player"),
            CountryProvider(id: 541, name: "rtve"),
            CountryProvider(id: 1773, name: "SkyShowtime"),
            CountryProvider(id: 337, name: "Disney Plus"),
            CountryProvider(id: 63, name: "Filmin"),
        ]),
        CountryCatalogEntry(regionCode: "IT", displayName: "Italy", originalLanguage: "it", providers: [
            CountryProvider(id: 8, name: "Netflix"),
            CountryProvider(id: 119, name: "Amazon Prime Video"),
            CountryProvider(id: 222, name: "Rai Play"),
            CountryProvider(id: 359, name: "Mediaset Infinity"),
            CountryProvider(id: 39, name: "Now TV"),
            CountryProvider(id: 109, name: "Timvision"),
            CountryProvider(id: 337, name: "Disney Plus"),
            CountryProvider(id: 29, name: "Sky Go"),
        ]),
        CountryCatalogEntry(regionCode: "IN", displayName: "India", originalLanguage: "hi", providers: [
            CountryProvider(id: 8, name: "Netflix"),
            CountryProvider(id: 119, name: "Amazon Prime Video"),
            CountryProvider(id: 2336, name: "JioHotstar"),
            CountryProvider(id: 232, name: "Zee5"),
            CountryProvider(id: 237, name: "Sony Liv"),
            CountryProvider(id: 515, name: "MX Player"),
            CountryProvider(id: 532, name: "aha"),
        ]),
        CountryCatalogEntry(regionCode: "BR", displayName: "Brazil", originalLanguage: "pt", providers: [
            CountryProvider(id: 8, name: "Netflix"),
            CountryProvider(id: 119, name: "Amazon Prime Video"),
            CountryProvider(id: 307, name: "Globoplay"),
            CountryProvider(id: 337, name: "Disney Plus"),
            CountryProvider(id: 531, name: "Paramount Plus"),
            CountryProvider(id: 47, name: "Looke"),
        ]),
        CountryCatalogEntry(regionCode: "MX", displayName: "Mexico", originalLanguage: "es", providers: [
            CountryProvider(id: 8, name: "Netflix"),
            CountryProvider(id: 119, name: "Amazon Prime Video"),
            CountryProvider(id: 457, name: "VIX"),
            CountryProvider(id: 337, name: "Disney Plus"),
            CountryProvider(id: 531, name: "Paramount Plus"),
            CountryProvider(id: 167, name: "Claro video"),
        ]),
        CountryCatalogEntry(regionCode: "AU", displayName: "Australia", originalLanguage: nil, providers: [
            CountryProvider(id: 8, name: "Netflix"),
            CountryProvider(id: 119, name: "Amazon Prime Video"),
            CountryProvider(id: 21, name: "Stan"),
            CountryProvider(id: 385, name: "BINGE"),
            CountryProvider(id: 135, name: "ABC iview"),
            CountryProvider(id: 132, name: "SBS On Demand"),
            CountryProvider(id: 378, name: "9Now"),
            CountryProvider(id: 337, name: "Disney Plus"),
        ]),
        CountryCatalogEntry(regionCode: "CA", displayName: "Canada", originalLanguage: nil, providers: [
            CountryProvider(id: 8, name: "Netflix"),
            CountryProvider(id: 119, name: "Amazon Prime Video"),
            CountryProvider(id: 230, name: "Crave"),
            CountryProvider(id: 337, name: "Disney Plus"),
            CountryProvider(id: 531, name: "Paramount Plus"),
            CountryProvider(id: 283, name: "Crunchyroll"),
        ]),
    ]

    /// Rotating destination of the day: `floor(now / 86_400_000) mod count`,
    /// computed at read time so the pick advances at UTC midnight and iOS and
    /// Android agree on the country for any given day. The index is always
    /// coerced non-negative.
    static var countryOfDay: CountryCatalogEntry {
        let count = entries.count
        let millis = Date().timeIntervalSince1970 * 1000
        let day = Int(floor(millis / 86_400_000))
        let idx = ((day % count) + count) % count
        return entries[idx]
    }

    /// Entry for a region code, case-insensitive; nil on miss.
    static func entry(forRegionCode code: String) -> CountryCatalogEntry? {
        entries.first { $0.regionCode.caseInsensitiveCompare(code) == .orderedSame }
    }
}

File: android/app/src/main/java/com/rork/guidestreamtvandroid/data/CountryCatalog.kt
package com.rork.guidestreamtvandroid.data

/**
 * One TMDB watch provider in a specific region.
 */
data class CountryProvider(
    val id: Int,
    val name: String,
    /** When set, this provider is queried in this language instead of the
     *  country's — lets Crunchyroll browse Japanese anime in every market. */
    val originalLanguage: String? = null,
)

/**
 * One curated "Around the World" destination. Mirrors iOS CountryCatalogEntry.
 */
data class CountryCatalogEntry(
    /** ISO 3166-1 alpha-2 region code (also TMDB's `watch_region` value). */
    val regionCode: String,
    val displayName: String,
    /** TMDB original-language filter, or null for English-speaking markets. */
    val originalLanguage: String?,
    /** Ordered TMDB provider entries — region-specific ids. */
    val providers: List<CountryProvider>,
) {
    /** Effective original language for a query against this provider — the
     *  provider override when present, else the country's. */
    fun effectiveOriginalLanguage(provider: CountryProvider): String? =
        provider.originalLanguage ?: originalLanguage

    /**
     * Vote-count floor for a query. 100 for Japanese content — a
     * content-safety requirement for animation: TMDB reports the titles it
     * excludes as adult:false with genre [16], so no other filter catches
     * them. 20 everywhere else. Do not lower it or make it uniform.
     */
    fun voteFloor(provider: CountryProvider): Int =
        if (effectiveOriginalLanguage(provider) == "ja") 100 else 20
}

/**
 * Curated "Around the World" destinations. TMDB provider ids are
 * region-specific — always use them exactly as listed. Mirrors
 * ios/Shared/CountryCatalog.swift, and must stay in the SAME ORDER as it:
 * both platforms derive the daily pick from the list position, so a reorder
 * on one side makes them disagree.
 *
 * "Around the World" means AWAY FROM HOME, and home is the viewer's own
 * region — not the US. Until 15 Sep 2026 there was no US entry and the
 * rotation was a global `day mod 11`, so a viewer in Tokyo was shown "Around
 * the World: Japan" one day in eleven, and nobody outside the US could ever
 * look at the largest catalogue on earth. Read the destination list through
 * [destinations], never [entries] directly.
 */
object CountryCatalog {
    val entries: List<CountryCatalogEntry> = listOf(
        CountryCatalogEntry("JP", "Japan", "ja", listOf(
            CountryProvider(84, "U-NEXT"),
            CountryProvider(8, "Netflix"),
            CountryProvider(9, "Prime Video"),
            CountryProvider(15, "Hulu"),
            CountryProvider(337, "Disney+"),
        )),
        CountryCatalogEntry("KR", "South Korea", "ko", listOf(
            CountryProvider(8, "Netflix"),
            CountryProvider(1883, "TVING"),
            CountryProvider(356, "wavve"),
            CountryProvider(97, "Watcha"),
            CountryProvider(283, "Crunchyroll", originalLanguage = "ja"),
        )),
        CountryCatalogEntry("GB", "United Kingdom", null, listOf(
            CountryProvider(38, "BBC iPlayer"),
            CountryProvider(41, "ITVX"),
            CountryProvider(103, "Channel 4"),
            CountryProvider(29, "Sky Go"),
            CountryProvider(39, "Now TV"),
            CountryProvider(283, "Crunchyroll", originalLanguage = "ja"),
        )),
        CountryCatalogEntry("FR", "France", "fr", listOf(
            CountryProvider(8, "Netflix"),
            CountryProvider(119, "Prime Video"),
            CountryProvider(337, "Disney+"),
            CountryProvider(283, "Crunchyroll", originalLanguage = "ja"),
        )),
        CountryCatalogEntry("DE", "Germany", "de", listOf(
            CountryProvider(8, "Netflix"),
            CountryProvider(9, "Prime Video"),
            CountryProvider(2750, "RTL+"),
            CountryProvider(337, "Disney+"),
            CountryProvider(30, "WOW"),
            CountryProvider(29, "Sky Go"),
            CountryProvider(283, "Crunchyroll", originalLanguage = "ja"),
        )),
        CountryCatalogEntry("ES", "Spain", "es", listOf(
            CountryProvider(8, "Netflix"),
            CountryProvider(62, "Atres Player"),
            CountryProvider(119, "Prime Video"),
            CountryProvider(2241, "Movistar+"),
            CountryProvider(337, "Disney+"),
            CountryProvider(541, "rtve"),
            CountryProvider(1773, "SkyShowtime"),
            CountryProvider(283, "Crunchyroll", originalLanguage = "ja"),
        )),
        CountryCatalogEntry("IT", "Italy", "it", listOf(
            CountryProvider(8, "Netflix"),
            CountryProvider(222, "Rai Play"),
            CountryProvider(119, "Prime Video"),
            CountryProvider(39, "Now TV"),
            CountryProvider(29, "Sky Go"),
            CountryProvider(337, "Disney+"),
            CountryProvider(283, "Crunchyroll", originalLanguage = "ja"),
        )),
        CountryCatalogEntry("IN", "India", "hi", listOf(
            CountryProvider(8, "Netflix"),
            CountryProvider(119, "Prime Video"),
            CountryProvider(2336, "JioHotstar"),
            CountryProvider(283, "Crunchyroll", originalLanguage = "ja"),
        )),
        CountryCatalogEntry("BR", "Brazil", "pt", listOf(
            CountryProvider(307, "Globoplay"),
            CountryProvider(8, "Netflix"),
            CountryProvider(119, "Prime Video"),
            CountryProvider(283, "Crunchyroll", originalLanguage = "ja"),
        )),
        CountryCatalogEntry("MX", "Mexico", "es", listOf(
            CountryProvider(8, "Netflix"),
            CountryProvider(457, "VIX"),
            CountryProvider(119, "Prime Video"),
            CountryProvider(337, "Disney+"),
            CountryProvider(167, "Claro video"),
            CountryProvider(283, "Crunchyroll", originalLanguage = "ja"),
        )),
        CountryCatalogEntry("AU", "Australia", null, listOf(
            CountryProvider(21, "Stan"),
            CountryProvider(385, "BINGE"),
            CountryProvider(378, "9Now"),
            CountryProvider(132, "SBS On Demand"),
            CountryProvider(135, "ABC iview"),
            CountryProvider(283, "Crunchyroll", originalLanguage = "ja"),
        )),
        // Added 15 Sep 2026. Every id below was read from TMDB's own US
        // provider list and then probed with /discover/tv?watch_region=US to
        // confirm it returns titles: Hulu 906, HBO Max 508, Peacock 471,
        // Paramount+ 354, Netflix 2315, Prime Video 1347, Disney+ 473,
        // Apple TV+ 153, Tubi 1272, Crunchyroll 1045. Hulu leads because the
        // US-only services are the point of visiting, the same reason GB
        // leads with BBC iPlayer and BR with Globoplay. US is LAST so a US
        // viewer's rotation is unchanged from before this entry existed.
        CountryCatalogEntry("US", "United States", null, listOf(
            CountryProvider(15, "Hulu"),
            CountryProvider(1899, "HBO Max"),
            CountryProvider(386, "Peacock"),
            CountryProvider(2303, "Paramount+"),
            CountryProvider(8, "Netflix"),
            CountryProvider(9, "Prime Video"),
            CountryProvider(337, "Disney+"),
            CountryProvider(350, "Apple TV+"),
            CountryProvider(73, "Tubi"),
            CountryProvider(283, "Crunchyroll", originalLanguage = "ja"),
        )),
    )

    /**
     * Where a viewer in [homeRegion] can travel: the catalogue minus their own
     * country. A viewer in Manila gets all twelve including the US; a viewer in
     * Tokyo gets eleven without Japan. Falls back to the full list if the
     * filter would empty it.
     *
     * This is the list the destination picker shows and the rail rotates
     * through. [entryFor] deliberately still searches everything, so a deep
     * link to your own country resolves rather than silently becoming
     * somewhere else.
     */
    fun destinations(homeRegion: String): List<CountryCatalogEntry> {
        val home = homeRegion.uppercase()
        val away = entries.filter { !it.regionCode.equals(home, ignoreCase = true) }
        return away.ifEmpty { entries }
    }

    /**
     * Rotating destination of the day for a viewer at [homeRegion]:
     * floor(now / 86_400_000) mod destinationCount, computed at read time so
     * the pick advances at UTC midnight. Android and iOS agree because they
     * build the same pool (this file's order, minus home) from the same day
     * index. The index is always coerced non-negative.
     */
    fun countryOfDay(homeRegion: String): CountryCatalogEntry {
        val pool = destinations(homeRegion)
        val count = pool.size
        val day = (System.currentTimeMillis() / 86_400_000L).toInt()
        val idx = ((day % count) + count) % count
        return pool[idx]
    }

    /** Today's destination for this device's own region. */
    val countryOfDay: CountryCatalogEntry
        get() = countryOfDay(DeviceLocale.region)

    /** Entry for a region code, case-insensitive; null on miss. */
    fun entryFor(regionCode: String): CountryCatalogEntry? =
        entries.firstOrNull { it.regionCode.equals(regionCode, ignoreCase = true) }
}
